lib/volcanic/authenticator/v1/http_request.rb

```diff
@@ -4,99 +4,69 @@
     #
-    # to extend class:
+    # example:
     #
-    # VaultRequest < Request
-    #   self.base_uri = 'http://abc.com/'
-    # end
+    #   HTTPRequest.get('/api/v1/user', body: {id: '1'}.to_json, headers: {'Content-Type' => '...'})
     #
-    # VaultRequest.get('api/v1/user')
+    #   extending the class:
     #
-    class HTTPRequest
+    #   VaultRequest < HTTPRequest
+    #     base_uri 'http://vault.cloud/'
+    #   end
+    #
+    #   VaultRequest.get('api/v1/user')
+    #
+    # This class also include +download+ method on Down gem.
+    # example:
+    #
+    #   HTTPRequest.download('/api/v1/image').read
+    #
+    #   # use default base_uri
+    #   KrakatoaRequest < HTTPRequest
+    #     base_uri 'http://krakatoa.cloud/'
+    #   end
+    #
+    #   KrakatoaRequest.download('/api/v1/image').read
+    #
+    # NOTE: only support +download+ for now.
+    #
+    class HTTPRequest < Base
+      include HTTParty
       class << self
-        attr_accessor :base_uri
-
-        def get(*args, &block)
-          RequestBase.new(base_uri, *args, &block).get
-        end
-
-        def post(*args, &block)
-          RequestBase.new(base_uri, *args, &block).post
-        end
-
-        def patch(*args, &block)
-          RequestBase.new(base_uri, *args, &block).patch
-        end
-
-        def put(*args, &block)
-          RequestBase.new(base_uri, *args, &block).put
-        end
-
-        def delete(*args, &block)
-          RequestBase.new(base_uri, *args, &block).delete
-        end
-
-        def download(*args)
-          RequestBase.new(base_uri).download(*args)
-        end
-      end
-
-      # request
-      class RequestBase
-        include HTTParty
-        extend Forwardable
-
-        attr_accessor :url, :args, :block
-        def_instance_delegator 'Volcanic::Authenticator.config'.to_sym, :auth_enabled?
-
-        def initialize(base_url, *args, &block)
-          @args = args
-          @block = block
-          @url = base_url
-          self.class.base_uri url
-          self.class.headers auth_header
-        end
-
-        def get
-          self.class.get(*args, &block)
-        end
-
-        def post
-          self.class.post(*args, &block)
-        end
-
-        def put
-          self.class.put(*args, &block)
-        end
-
-        def patch
-          self.class.patch(*args, &block)
+        # extend this class to have +download+ method from Down gem.
+        # +base_uri+ and +headers+ is coming from HTTParty module class methods.
+        def download(endpoint, **opts)
+          path = [base_uri, endpoint].join
+          opts[:headers] = headers
+          Down.download(path, opts)
         end
 
-        def delete
-          self.class.delete(*args, &block)
-        end
+        # Override HTTParty +headers+ method to enable dynamic
+        # authorization header.
+        def headers(headers = nil)
+          return super headers if headers
```

Comment:
so... if an additional header is needed, with this change it will be impossible to provide that header and add the auth header...

If you change the method signature to be `def headers(headers = {})` then you will always have a headers hash inside the method. This will allow you to do something like:
```
base_headers = auth_enabled? ? { 'Authorization' => auth_token } : {}
super base_headers.merge(headers)
```
This has the benefits that:

- the base class' implementation is always used in case it does something special
- the authorization header will be set if needed
- any headers passed in will be respected (including the authorization header if the caller wants to override that)

Reply:
this is much more simpler. thanks for the tips.